File: prob1.py
# Time Complexity :O(n)
# Space Complexity :O(h) - height of tree
# Leet Code: Yes

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        res = []
        def helper(root, arr, targetSum):
            if root is None:
                return
            arr.append(root.val)
            s = sum(arr)
            if root.left == None and root.right == None and s == targetSum:
                newArr = arr[:]
                res.append(newArr)
            
            helper(root.left, arr[:], targetSum)
            helper(root.right, arr[:], targetSum)
            #backtracking
            arr.pop()

        helper(root, [], targetSum)
        return res

File: prob1.java
// Time Complexity : O(n2)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes


class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if(root == null) return new ArrayList<>();
        return helper(root, 0, targetSum, new ArrayList<>(), new ArrayList<>());
    }
    List<List<Integer>> helper(TreeNode node, int sum, int target, List<List<Integer>> ans, List<Integer> list){
        if(node == null) return ans;
        list.add(node.val);
        if(node.left == null && node.right == null && sum + node.val == target){
            ans.add(new ArrayList<>(list));
        }
        helper(node.left, sum + node.val, target, ans, list);
        helper(node.right, sum + node.val, target, ans, list);
        list.remove(list.size() - 1);
        return ans;
    }
}